# Exploring eBay Car Sales Data

## Introduction

In this project we are going to analyze eBay car sales data. Our main goal in this project is to consolidate what we have learned in this course. We are going to work with `numpy` and `pandas` modules.

Throughout this project we are going to use various methods that will make working with our dataset easier. For example, after cleaning we will be left with 46,352 entries instead of 50,000.

## Opening the File

Before we start to open the file, we need to import modules that will help us in our analysis.

In [12]:
import numpy as np
import pandas as pd

Now we can open and read our file. We are going to use a dataset that was scraped and uploaded to [Kaggle](https://www.kaggle.com/) by user orgesleka. The original dataset isn't available on [Kaggle](https://www.kaggle.com/) anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

In [13]:
# read the dataset into dataframe using 'latin-1' encoding
autos = pd.read_csv('data/autos.csv', encoding='latin-1')

Here is our dataset table. Let's first find out what each column means.

In [14]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `odometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [15]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

And here is some info about our columns. We can see here that a few columns have empty values. For example `vehicleType` with 44,905 values and `fuelType` with 45,518. There are also many `object` type data. We will use this information later in our project.

## Changing the Columns

We already saw the columns of our dataset, but let's look at it one more time.

In [16]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

One thing we can notice is that the columns use `CamelCase` instead of `snake_case`. For convenience, we will change them to `snake_case`.

In [17]:
# changing CamelCase columns to snake_case
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']

autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

With the new columns, we can continue our analysis.

## Exploring Price and Odometer Columns

Let's take a look at the values in our dataset.

In [18]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In `price` and `odometer` columns, we have numeric data with some symbols. For `price` column we have dollar sign followed by a number. In `odometer` column, we have km alongside with number. We should change these columns into numeric data to do some calculations later.

First, we are going to drop the dollar sign in `price` column and leave only a number.

In [19]:
# converting `price` column from `object` to `int` datatype
autos['price'] = (autos['price']
                        .str.replace(',','')
                        .str.replace('$','')
                        .astype(int)
                 )

For `odometer` column we are going to drop km and leave only a number. We are also renaming `odometer` column to `odometer_in_km` for better understanding in the future.

In [20]:
# converting `odometer` column from `object` to `int` datatype
autos['odometer'] = (autos['odometer']
                            .str.replace('km','')
                            .str.replace(',','')
                            .astype(int)
                    )

# changing `odometer` column name to `odometer_in_km`
autos.rename({'odometer': 'odometer_in_km'}, axis=1, inplace=True)

This is what our columns look like after the changes.

In [21]:
autos[['price', 'odometer_in_km']].head()

price  odometer_in_km
0   5000          150000
1   8500          150000
2   8990           70000
3   4350           70000
4   1350          150000

### Removing Mispriced Entries

There are many mispriced entries in our dataset. Let's take a look at the lowest and the highest prices.

In [22]:
print(autos['price'].value_counts().sort_index(ascending=True).head(50))
print(autos['price'].value_counts().sort_index(ascending=False).head(15))

price
0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
70       10
75        5
79        1
80       15
89        1
90        5
99       19
100     134
110       3
111       2
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
Name: count, dtype: int64
price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: count, dtype: int64


For lowest prices, there are some entries like 0 and 1, we are going to drop them up to 100. For highest prices, there are some weird entries such as 999,990, 999,999 and 1,234,566, we are going to drop them too.

In [23]:
# dropping incorrect price data and leave only values between 100 and 350,000
autos = autos.loc[autos['price'].between(100,350000)]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48224 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48224 non-null  object
 1   name                48224 non-null  object
 2   seller              48224 non-null  object
 3   offer_type          48224 non-null  object
 4   price               48224 non-null  int64 
 5   abtest              48224 non-null  object
 6   vehicle_type        43801 non-null  object
 7   registration_year   48224 non-null  int64 
 8   gearbox             46019 non-null  object
 9   power_ps            48224 non-null  int64 
 10  model               45829 non-null  object
 11  odometer_in_km      48224 non-null  int64 
 12  registration_month  48224 non-null  int64 
 13  fuel_type           44345 non-null  object
 14  brand               48224 non-null  object
 15  unrepaired_damage   39340 non-null  object
 16  ad_created          48224 n

At this point, we have cleared the `price` column and also changed the data type to numerical for `odometer_in_km` and `price` columns.

## Exploring Date Columns

Let's look at some date columns in our dataset.

In [24]:
# printing date only without time, sorted by index
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

date_crawled
2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: proportion, dtype: float64


We can see that the data for this dataset was collected from 2016/03/05 up to 2016/04/07. But there's nothing interesting.

In [25]:
# printing date only without time, sorted by index
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038860
2016-04-04    0.036890
2016-04-05    0.011799
2016-04-06    0.003256
2016-04-07    0.001244
Name: proportion, Length: 76, dtype: float64


Based on this column we can say that our dataset contains ads from 2015/06/11 up to 2016/04/07. 

In [26]:
# printing date only without time, sorted by index
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

last_seen
2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010638
2016-03-11    0.012400
2016-03-12    0.023785
2016-03-13    0.008875
2016-03-14    0.012629
2016-03-15    0.015863
2016-03-16    0.016444
2016-03-17    0.028098
2016-03-18    0.007320
2016-03-19    0.015760
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-22    0.021359
2016-03-23    0.018580
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-26    0.016672
2016-03-27    0.015552
2016-03-28    0.020840
2016-03-29    0.022292
2016-03-30    0.024697
2016-03-31    0.023826
2016-04-01    0.022852
2016-04-02    0.024884
2016-04-03    0.025133
2016-04-04    0.024531
2016-04-05    0.125062
2016-04-06    0.221964
2016-04-07    0.132154
Name: proportion, dtype: float64


As for `last_seen` column, we can say that the values pretty the same as for `date_crawled` column, from 2016/03/05 up to 2016/04/07.

### Removing Wrong Data from Registration Year Column

If we look at lowest and highest values of `registration_year` column, we can see some wrong data. 

In [27]:
print(autos['registration_year'].value_counts().sort_index().head(5))
print(autos['registration_year'].value_counts().sort_index().tail(15))

registration_year
1000    1
1001    1
1111    1
1800    2
1910    2
Name: count, dtype: int64
registration_year
2015     380
2016    1202
2017    1383
2018     468
2019       2
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: count, dtype: int64


There are years such as 1000, 1001, 2800 and 5000. As our dataset was scraped in 2016 we can assume that the last year of registration should be in 2016. As for lowest year, it could be somewhere in the first few decades of the 1900s.

In [28]:
# dropping incorrect `registration_year` data and leave only values between 1910 and 2016
autos = autos.loc[autos['registration_year'].between(1910,2016)]
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
Index: 46352 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        46352 non-null  object
 1   name                46352 non-null  object
 2   seller              46352 non-null  object
 3   offer_type          46352 non-null  object
 4   price               46352 non-null  int64 
 5   abtest              46352 non-null  object
 6   vehicle_type        43799 non-null  object
 7   registration_year   46352 non-null  int64 
 8   gearbox             44378 non-null  object
 9   power_ps            46352 non-null  int64 
 10  model               44221 non-null  object
 11  odometer_in_km      46352 non-null  int64 
 12  registration_month  46352 non-null  int64 
 13  fuel_type           43181 non-null  object
 14  brand               46352 non-null  object
 15  unrepaired_damage   38254 non-null  object
 16  ad_created          46352 n

Before we deleted the incorrect data, we had 48,224 values, and after cleaning we have 46,352. Let's find out in which year the largest number of cars was registered.

In [29]:
print(autos['registration_year'].value_counts(normalize=True))

registration_year
2000    0.066966
2005    0.062802
1999    0.062112
2004    0.058228
2003    0.058099
          ...   
1939    0.000022
1948    0.000022
1938    0.000022
1953    0.000022
1950    0.000022
Name: proportion, Length: 78, dtype: float64


Of all our dataset, 6% of the cars were registered in 2000, 2005 and 1999. For 2004 and 2003 this is 5%.

## Exploring Brand Column

Let's take a look at `brand` column and pick some brands for our next steps.

In [30]:
print(autos['brand'].value_counts(normalize=True))

brand
volkswagen        0.211404
bmw               0.110179
opel              0.107245
mercedes_benz     0.096652
audi              0.086771
ford              0.069835
renault           0.047075
peugeot           0.029858
fiat              0.025608
seat              0.018252
skoda             0.016418
nissan            0.015339
mazda             0.015231
smart             0.014196
citroen           0.014045
toyota            0.012793
hyundai           0.010010
sonstige_autos    0.009536
volvo             0.009126
mini              0.008802
mitsubishi        0.008177
honda             0.007875
kia               0.007076
alfa_romeo        0.006666
porsche           0.006019
suzuki            0.005933
chevrolet         0.005674
chrysler          0.003517
dacia             0.002654
daihatsu          0.002503
jeep              0.002287
land_rover        0.002114
subaru            0.002114
saab              0.001661
jaguar            0.001532
daewoo            0.001489
trabant           0.00

We are going to select first 6 brands that have over 5% of the all cars in our dataset. The first step is to find the average price for each brand of cars.

In [31]:
# counting mean price of cars by brand
mean_price_by_brand = dict()

for brand in autos['brand'].value_counts().head(6).index:
    
    mean_price = round(autos.loc[autos['brand'] == brand, 'price'].mean())
    
    mean_price_by_brand[brand] = mean_price

print(mean_price_by_brand)

{'volkswagen': 5437, 'bmw': 8382, 'opel': 3005, 'mercedes_benz': 8673, 'audi': 9381, 'ford': 3779}


As we can see, 'Audi', 'Mercedes-Benz' and 'BMW' have the highest price among the selected brands. 'Ford' and 'Opel' have the lowest prices and 'Volkwagen' somewhere in the middle. Let's find out if the mileage affects the price somehow.

In [32]:
# counting mean mileage of cars by brand
mean_mileage_by_brand = dict()

for brand in autos['brand'].value_counts().head(6).index:
    
    mean_mileage = round(autos.loc[autos['brand'] == brand, 'odometer_in_km'].mean())
    
    mean_mileage_by_brand[brand] = mean_mileage

print(mean_mileage_by_brand)

{'volkswagen': 128800, 'bmw': 132695, 'opel': 129384, 'mercedes_benz': 131026, 'audi': 129245, 'ford': 124277}


Now we have the average price and the average mileage. Let's compare them and make some conclusions.

In [33]:
# creating `mean_price` and `mean_mileage` dataframe to compare them
bmp_series = pd.Series(mean_price_by_brand)
bmm_series = pd.Series(mean_mileage_by_brand)

price_mileage_df = pd.DataFrame(bmp_series, columns=['mean_price'])

price_mileage_df['mean_mileage'] = bmm_series

price_mileage_df

mean_price  mean_mileage
volkswagen           5437        128800
bmw                  8382        132695
opel                 3005        129384
mercedes_benz        8673        131026
audi                 9381        129245
ford                 3779        124277

In the table above we can see that 'BMW' have average price \\$8,382 and average mileage 132,695km. As for 'Ford' average price is \\$3,779 and average mileage 124,277km. We can say that mileage doesn't affect the price in this situation.

## Extra Steps

### Changing German Words into English

There are also many words written in German. We can find some of them in our full dataset table.

In [34]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_in_km  registration_month  \
0        manuell       158  andere          150000                   3   
1      automatik       286     7er          150000                   6   
2        manuell       102    golf           70000                   7   
3      automatik        71  fortwo           70000                   6   
4        manuell         0   focus          150000                   7   
...          ...       ...     ...             ...                 ...   
49995  automatik       239      q5          100000                   1   
49996    manuell        75   astra          150000                   5   
49997  automatik        69     500            5000                  11   
49998    manuell       150      a3           40000                  11   
49999    manuell       101  vectra          150000                   1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       number_of_pictures  postal_code            last_seen  
0                       0        79588  2016-04-06 06:45:54  
1                       0        71034  2016-04-06 14:45:08  
2                       0        35394  2016-04-06 20:15:37  
3                       0        33729  2016-03-15 03:16:28  
4                       0        39218  2016-04-01 14:38:50  
...                   ... 

There are many columns that contain German words. We will print all of them below.

In [35]:
print(autos['seller'].unique())
print(autos['offer_type'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['privat' 'gewerblich']
['Angebot']
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


We are going to translate these words written in German and swap them with English translation.

In [36]:
# creating maps for columns that contains German words
autos['seller'] = autos['seller'].map({
                                       'privat': 'private',
                                       'gewerblich': 'commercial'
                                     })

autos['offer_type'] = autos['offer_type'].map({'Angebot': 'offer'})

vehicle_type_map = {
                    'bus': 'bus',
                    'limousine': 'limousine',
                    'kleinwagen': 'small car',
                    'kombi': 'station wagon',
                    'coupe': 'coupe',
                    'suv': 'suv',
                    'cabrio': 'cabrio',
                    'andere': 'other'
                    }
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type_map)

autos['gearbox'] = autos['gearbox'].map({
                                         'manuell': 'manual',
                                         'automatik': 'automatic'
                                         })

fuel_type_map = {
                 'lpg': 'lpg',
                 'benzin': 'gasoline',
                 'diesel': 'diesel',
                 'cng': 'cng',
                 'hybrid': 'hybrid',
                 'elektro': 'electro',
                 'andere': 'other'
                 }
autos['fuel_type'] = autos['fuel_type'].map(fuel_type_map)

autos['unrepaired_damage'] = autos['unrepaired_damage'].map({
                                                             'nein': 'no',
                                                             'ja': 'yes'
                                                             })

In [37]:
print(autos['seller'].unique())
print(autos['offer_type'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['private' 'commercial']
['offer']
['bus' 'limousine' 'small car' 'station wagon' nan 'coupe' 'suv' 'cabrio'
 'other']
['manual' 'automatic' nan]
['lpg' 'gasoline' 'diesel' nan 'cng' 'hybrid' 'electro' 'other']
['no' nan 'yes']


Here are the translated versions of words.

### Changing Date into Numerical Data Type

There is also date columns with the following format "YYYY-MM-DD" alongside with time values. We don't need exact time when data was crawled or when ad was posted so we can drop time part and leave only date.

In [38]:
autos[['date_crawled', 'ad_created', 'last_seen']]

date_crawled           ad_created            last_seen
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
...                    ...                  ...                  ...
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48

[46352 rows x 3 columns]

We are also converting our date into the numeric data type.

In [39]:
def date_to_numerical(df, date_column):
    numerical_date = (df[date_column]
                         .str.split()
                         .str[0]
                         .str.replace('-','')
                         .astype(int)
                     )
    return numerical_date

In [40]:
autos['date_crawled'] = date_to_numerical(autos, 'date_crawled')

autos['ad_created'] = date_to_numerical(autos, 'ad_created')

autos['last_seen'] = date_to_numerical(autos, 'last_seen')

You can see the result of our work down below.

In [41]:
autos[['date_crawled', 'ad_created', 'last_seen']]

date_crawled  ad_created  last_seen
0          20160326    20160326   20160406
1          20160404    20160404   20160406
2          20160326    20160326   20160406
3          20160312    20160312   20160315
4          20160401    20160401   20160401
...             ...         ...        ...
49995      20160327    20160327   20160401
49996      20160328    20160328   20160402
49997      20160402    20160402   20160404
49998      20160308    20160308   20160405
49999      20160314    20160313   20160406

[46352 rows x 3 columns]

### Finding the Most Common Brand/Model Combination

For our next step we are going to find out which brand/model combination is most common in our dataset.

In [42]:
brand_model = autos['brand'] + ' ' + autos['model']

print(f'The {brand_model.describe()[2].title()} is the most common brand/model combination')

The Volkswagen Golf is the most common brand/model combination


Look's like "Volkswagen Golf" is the most common brand/model combination.

### Finding How Mileage Affects the Average Price of Auto

Now, let's find out how mileage affects the price of cars. First, let's look at the lowest and highest mileage.

In [43]:
lowest_mileage = autos['odometer_in_km'].min()
highest_mileage = autos['odometer_in_km'].max()

print(f'The lowest mileage in dataset is {lowest_mileage:,}km', )
print(f'The highest mileage in dataset is {highest_mileage:,}km', )

The lowest mileage in dataset is 5,000km
The highest mileage in dataset is 150,000km


The lowest mileage in our dataset is 5,000km and the highest is 150,000km. Let's split values into three groups. From 5,000km to 50,000km, from 50,000km to 100,000km and from 100,000km up to 150,000km.

In [44]:
odometer_5_to_50 = autos.loc[autos['odometer_in_km'].between(5000, 50000), 'price'].mean()
odometer_50_to_100 = autos.loc[autos['odometer_in_km'].between(50001, 100000), 'price'].mean()
odometer_100_to_150 = autos.loc[autos['odometer_in_km'].between(100001, 150000), 'price'].mean()

print(
      'The mean price of the cars with mileage from 5,000km up to 50,000km '
      f'is ${round(odometer_5_to_50):,}'
     )
print(
      'The mean price of the cars with mileage from 50,000km up to 100,000km '
      f'is ${round(odometer_50_to_100):,}'
     )
print(
      'The mean price of the cars with mileage from 100,000km up to 150,000km '
      f'is ${round(odometer_100_to_150):,}'
     )

The mean price of the cars with mileage from 5,000km up to 50,000km is $15,214
The mean price of the cars with mileage from 50,000km up to 100,000km is $9,625
The mean price of the cars with mileage from 100,000km up to 150,000km is $4,133


The mean car price in the first group is \\$15,214. For the second group, the mean price is \\$9,625. And for the last group with the highest mileage the mean price is \\$4,133. We can say that the lower the mileage the higher the price.

### Finding How Unrepaired Damage Affects the Average Price of Auto

For our last step, we are going to find out how unrepaired damage affects the mean price of the cars.

In [45]:
damaged_mean_price = autos.loc[autos['unrepaired_damage'] == 'yes', 'price'].mean()
undamaged_mean_price = autos.loc[autos['unrepaired_damage'] == 'no', 'price'].mean()

print(
      'The mean price of the cars with unrepaired damage '
      f'is ${round(damaged_mean_price):,}'
     )
print(
      'The mean price of the cars without unrepaired damage '
      f'is ${round(undamaged_mean_price):,}'
     )

The mean price of the cars with unrepaired damage is $2,270
The mean price of the cars without unrepaired damage is $7,177


As we can see, the unrepaired damage heavily affects the mean price of cars. For the cars with unrepaired damage the mean price is \\$2,270. And for the cars without unrepaired damage the mean price is \\$7,177.

## Conclusion

In the beginning, our goal was to consolidate our skills that we have learned during this course. Throughout this project we have used some techniques to make our dataset cleaner and easier to read.

Also, during our analysis we find out that the mean price of the auto is heavily affected by unrepaired damage and high mileage. The cars with unrepaired damage or high mileage have a very low price compared to undamaged or low mileage cars.